compiler/src/middle_end/matchcomp.re

```diff
@@ -719,120 +368,200 @@ let rec rotate_if_needed = mtx =>
+    switch (row) {
+    | [] => failwith("Impossible: Empty pattern row in specialize_matrix")
+    | [{pat_desc} as p, ...ptl] =>
+      switch (pat_desc) {
+      | TPatAlias(p, alias, _) =>
+        specialized_rows([p, ...ptl], [(alias, cur), ...binds])
+      | TPatVar(id, _) =>
+        let wildcard = Parmatch.omega;
+        [([wildcard, ...ptl], [(id, cur), ...binds])];
+      | TPatConstant(c) when c == const =>
+        let wildcard = Parmatch.omega;
+        [([wildcard, ...ptl], binds)];
+      | TPatOr(p1, p2) =>
+        specialized_rows([p1, ...ptl], binds)
+        @ specialized_rows([p2, ...ptl], binds)
+      | _ when pattern_always_matches(p) => [(row, binds)]
+      | _ => [] /* Specialization for non-constants generates no rows. */
+      }
+    };
+
+  List.fold_right(
+    ((row, binds, mb), rem) =>
+      List.map(
+        ((patts, binds)) => (patts, binds, mb),
+        specialized_rows(row, binds),
+      )
+      @ rem,
+    mtx,
+    [],
+  );
+};
+
+/* [See paper for details; modified slightly]
 
       Row: ([p_1; p_2; ...], a_j)
 
       Pattern p_1 (for row j)       Row(s) of D(P)
       -----------------------       --------------
            c(q_1,...,q_n)               No row
-                 _                [([p_2; ...], a_j)]
+                 _                [([p_1; p_2; ...], a_j)]
             (q_1 | q_2)           [D(([q_1; p_2; ...], a_j));
                                    D(([q_2; p_2; ...], a_j))]
    */
-let rec default_matrix = mtx => {
-  let rec default_rows = row =>
+let rec default_matrix = (cur, mtx) => {
+  let rec default_rows = (row, binds) =>
     switch (row) {
     | [] => failwith("Impossible: Empty pattern row in default_matrix")
     | [{pat_desc} as p, ...ptl] =>
       switch (pat_desc) {
-      | _ when pattern_always_matches(p) => [ptl]
+      | TPatAlias(inner, alias, _) =>
+        default_rows([inner, ...ptl], [(alias, cur), ...binds])
+      | _ when pattern_always_matches(p) => [(row, binds)]
       | TPatOr(p1, p2) =>
-        default_rows([p1, ...ptl]) @ default_rows([p2, ...ptl])
+        default_rows([p1, ...ptl], binds)
+        @ default_rows([p2, ...ptl], binds)
       | _ => []
       }
     };
 
   List.fold_right(
-    ((row, mb), rem) =>
-      List.map(patts => (patts, mb), default_rows(row)) @ rem,
+    ((row, binds, mb), rem) =>
+      List.map(
+        ((patts, binds)) => (patts, binds, mb),
+        default_rows(row, binds),
+      )
+      @ rem,
     mtx,
     [],
   );
 };
 
+let equality_type =
+  fun
+  | Const_number(_)
+  | Const_int32(_)
+  | Const_int64(_)
+  | Const_float32(_)
+  | Const_float64(_)
+  | Const_bytes(_)
+  | Const_string(_) => StructuralEquality
+  | Const_void
+  | Const_bool(_)
+  | Const_char(_)
+  | Const_wasmi32(_) => PhysicalEquality
+  | Const_wasmi64(_)
+  | Const_wasmf32(_)
+  | Const_wasmf64(_) =>
+    failwith(
+      "Pattern matching not supported on low-level i64/f32/f64 types.",
+    );
```

Comment:
With these pattern matching updates, how far are we from supporting these?

Reply:
"this is only really possible with wasm GC" far haha